pkg/controller/linstornodeset/linstornodeset_controller.go

```diff
@@ -329,180 +333,260 @@ func (r *ReconcileLinstorNodeSet) reconcileResource(ctx context.Context, pns *pi
-		err error
-	}
-
-	satelliteStatusIn := make(chan satStat)
-	satelliteStatusOut := make(chan error)
-
-	maxConcurrentNodes := 5
-	tokens := make(chan struct{}, maxConcurrentNodes)
-
-	var wg sync.WaitGroup
-
-	pns.Status.SatelliteStatuses = make([]*piraeusv1alpha1.SatelliteStatus, len(pods.Items))
-
-	for i := range pods.Items {
-		wg.Add(1)
-		pod := pods.Items[i]
-
-		log := logrus.WithFields(logrus.Fields{
-			"podName":                        pod.Name,
-			"podNameSpace":                   pod.Namespace,
-			"podPhase":                       pod.Status.Phase,
-			"podNumber":                      i,
-			"maxConcurrentNodeRegistrations": maxConcurrentNodes,
-		})
-		log.Debug("NS reconcileSatNodes: reconciling node")
-
-		sat := &piraeusv1alpha1.SatelliteStatus{
-			NodeStatus:          piraeusv1alpha1.NodeStatus{NodeName: pod.Spec.NodeName},
-			StoragePoolStatuses: make([]*piraeusv1alpha1.StoragePoolStatus, 0),
-		}
-		pns.Status.SatelliteStatuses[i] = sat
-
+	output := make(chan error)
+	for _, item := range pods {
+		pod := item
+		// Registration can be done in parallel, so we handle per-node work in a separate go-routine
 		go func() {
-			l := log
-			l.Debug("NS reconcileSatNodes: waiting to acquire token...")
-			tokens <- struct{}{} // Acquire a token
-			l.Debug("NS reconcileSatNodes: token acquired, registering node")
-
-			err := r.reconcileSatNodeWithController(sat, pod, pns.Spec.SslConfig)
+			log := logrus.WithFields(logrus.Fields{
+				"podName":      pod.Name,
+				"podNameSpace": pod.Namespace,
+				"podPhase":     pod.Status.Phase,
+			})
+
+			log.Debug("reconcile node registration")
+			err := r.reconcileSingleNodeRegistration(ctx, nodeSet, pod)
 			if err != nil {
-				l.Debugf("NS reconcileSatNodes: Error with reconcileSatNodeWithController: %v", err)
+				output <- err
+				return
 			}
-			satelliteStatusIn <- satStat{sat, err}
-		}()
-
-		pools := r.aggregateStoragePools(pns)
 
-		go func() {
-			l := log
-			defer func() {
-				<-tokens // Work done, release token.
-				l.Debug("NS reconcileSatNodes: token released")
-				wg.Done()
-			}()
-
-			in := <-satelliteStatusIn
-			if in.err != nil {
-				satelliteStatusOut <- in.err
+			log.Debug("reconcile automatic device setup")
+			err = r.reconcileAutomaticDeviceSetup(ctx, nodeSet, pod)
+			if err != nil {
+				output <- err
 				return
 			}
 
-			err := r.reconcileStoragePoolsOnNode(in.sat, pools, pod)
+			log.Debug("reconcile storage pool setup")
+			err = r.reconcileStoragePoolsOnNode(ctx, nodeSet, pod)
 			if err != nil {
-				l.Debug("NS reconcileSatNodes: Error with reconcileStoragePoolsOnNode")
+				output <- err
+				return
 			}
-			satelliteStatusOut <- err
+
+			output <- nil
 		}()
 	}
 
-	go func() {
-		wg.Wait()
-		close(satelliteStatusOut)
-	}()
-
-	var compoundError []error
-
-	for err := range satelliteStatusOut {
+	logger.Debug("start collecting per-node reconciliation results")
+	registerErrors := make([]error, 0)
+	for range pods {
```

Comment:
Maybe worth adding a comment here that we are collecting the number of outputs equal to the number of pods, rather than precisely the outputs for each pod. Otherwise someone will later add logic depending on the pod and they will be mixed up.